File: imagine/crop/__crop.py
from dotwiz import DotWiz

from imagine.core import global_helpers
from imagine.crop.crop_category import CropCategory
from imagine.crop.growth_model import GrowthModel
from imagine.events.imagine_event import ImagineEvent
from imagine.core.price_model import PriceModel
from imagine.util.absorb_fields import absorb_fields
from imagine.crop.growth_model_registry import *
from imagine.util.numpify_dictionary import numpify_dictionary
import numpy as np


print("loading __crop.py")
class Crop:
    def __init__(self, d=None):

        self.name = ""
        self.colour = [0, 0, 0]
        self.financial_events = []
        self.growth_model = None
        self.category = None
        # Changing price models implementation. Now using 'price_models'.
        # Price models will be included in lists applicable for each event that the price.
        # For Products,
        # self.product_price_models = []
        # self.cost_price_models = []
        self.price_config = None
        self.crop_rate_defs = None
        self.all_events = None

        if isinstance(d, dict):
            self.load_crop(d)
        else:
            self._setup_events()
            self._initialise_price_config()

#        self.colour = colour_in
#        self.financial_events = financial_events_in if financial_events_in is not None else [ImagineEvent.empty(1, 0)]
#        self.category_index = 0
#        self.categories = CropCategory.setup_categories()
#         if gm_in is not None:
#             self.growth_model = gm_in
 #       if category_in is not None:
 #           self.category = category_in

    # @staticmethod
    # def loadobj(obj):
    #     obj.category_choice = obj.category_choice
    #     obj.growth_model.name

    # @property
    # def category(self):
    #     if self.category_index:
    #         return self.categories[self.category_index]
    #     else:
    #         return CropCategory.empty(1, 0)

    # @category.setter
    # def category(self, cat):
    #     ix = next((i for i, c in enumerate(self.categories) if c.name == cat.name), None)
    #     if ix is None:
    #         self.categories.append(cat)
    #         self.category_index = len(self.categories) - 1
    #     else:
    #         self.categories[self.category_index] = cat
    #
    # def get_saved_category(self, cat_name):
    #     ix = next((i for i, c in enumerate(self.categories) if c.name == cat_name), None)
    #     if ix is None:
    #         return []
    #     else:
    #         return self.categories[ix]

    def get_cost_price_models(self, event_name):
        # Use the crop_rate_defs to look up the indices that map
        # the event's price models into the crop price models list.
        if event_name in self.price_config.cost_price_mappings:
            relevant_indices = self.price_config.cost_price_mappings[event_name]
            return self.price_config.cost_price_models[relevant_indices]
        return []

    @property
    def category_choice(self):
        if self.category:
            return self.category.name
        return ""

        # if self.category_index:
        #     return self.category.name
        # else:
        #     return ''

    # @category_choice.setter
    # def category_choice(self, cat_choice):
    #     cat = CropCategory(cat_choice)
    #     if cat.is_empty():
    #         print('Unable to change categories in crop. Supplied category is not a valid choice.')
    #         return
    #     ix = next((i for i, c in enumerate(self.categories) if c.name == cat_choice), None)
    #     if ix is not None:
    #         self.category_index = ix
    #     else:
    #         self.category = cat

    # @property
    # def growth_model(self):
    #     return self.category.growth_model

    def get_events(self):
        return self.all_events

    def get_unique_price_model_definitions(self, type_):
        pms = []
        if type_ == 'Income':
            price_models = self.growth_model.product_price_models
        elif type_ == 'Cost':
            gmes = self.growth_model.growth_model_events
            cost_price_models = [gme.cost_price_model for gme in gmes]
            financial_price_models = [event.cost_price_model for event in self.financial_events]
            price_models = cost_price_models + financial_price_models
        else:
            raise ValueError('Must pass a type argument. Valid types are ''Income'' and ''Cost''.')

        for pm in price_models:
            if not any(pm.definition_matches(p) for p in pms):
                pms.append(pm)

        return pms

    # @staticmethod
    # def absorb(orig_obj, new_obj):
    #     if type(new_obj) == type(orig_obj):
    #         for prop_name in orig_obj.__dict__.keys():
    #             if not prop_name.startswith('_'):
    #                 setattr(orig_obj, prop_name, getattr(new_obj, prop_name))
    #     else:
    #         raise ValueError('Absorb method can only be called on objects of exactly the same type.')

    # @staticmethod
    # def is_valid(crops):
    #     valid = []
    #     for c in crops:
    #         valid

    # Loads general crop information from the dictionary d into the object self.
    def load_crop(self, d):
        if not "name" in d:
            raise ValueError(f"Crop config does not define a name.")

        # Simple fields to load from config.
        simple_fields = [
            'name',
            'colour',
        ]
        absorb_fields(self, d, simple_fields)

        # Load category
        self.category = CropCategory.get_category(d.category_name)

        # Load growth model
        self.growth_model = globals()[d.growth_model_name](self.category, d.growth_model_parameters)

        # Load financial events
        if 'financial_events' in d:
            self.financial_events = [ImagineEvent.load_event(fe) for fe in d.financial_events]

        self._setup_events()
        self._initialise_price_config()
        self._setup_price_config(d)
        self._setup_crop_rate_defs()
        self._build_product_price_mappings()
        # Convert lists in price_config to numpy arrays to provide list based slicing.
        numpify_dictionary(self.price_config)

    def _setup_events(self):

        self.all_events = DotWiz()
        self.all_events.initial_events = self.growth_model.growth_model_initial_events
        self.all_events.regular_events = self.growth_model.growth_model_regular_events
        self.all_events.destruction_events = self.growth_model.growth_model_destruction_events
        self.all_events.gm_financial_events = self.growth_model.growth_model_financial_events
        self.all_events.crop_financial_events = self.financial_events

    # Puts all the columns of units from various growth model outputs into a handy structure we can just look up.
    # Must be called after _setup_events()
    def _initialise_price_config(self):
        self.price_config = DotWiz({
            'product_price_models': [],
            'cost_price_models': [],
            'relevant_product_price_model_indices': {
                'generic': [],
                'default': []
            },
            'cost_price_mappings': {},
            'product_price_mappings': {}
        })
        for event_type, events in self.all_events.items():
            for event in events:
                self.price_config.relevant_product_price_model_indices[event.name] = []
                self.price_config.cost_price_mappings[event.name] = []
                self.product_price_mappings = []

    def _setup_price_config(self, d):
        if 'product_price_models' not in d:
            raise ValueError(f"No product_price_models defined in crop config for {self.name}.")
        for pm_ix, pm_dict in enumerate(d.product_price_models):
            # First part of building the config is to create the list of all price models.
            if "name" not in pm_dict:
                # Add a product name if it's missing. Doesn't apply to costs. Costs require a name. (and event_name(s))
                pm_dict.name = pm_dict.rate.unit.measurable + " Income"
            self.price_config.product_price_models.append(PriceModel(pm_dict))

            # Next we have to record which price models are relevant to which events (via their indices).
            event_names = []
            if "event_name" in pm_dict:
                event_names = [pm_dict.event_name]
            elif "event_names" in pm_dict:
                event_names = pm_dict.event_names

            if event_names:
                for event_name in event_names:
                    if event_name in self.price_config.relevant_product_price_model_indices:
                        self.price_config.relevant_product_price_model_indices[event_name].append(pm_ix)
            else:
                self.price_config.relevant_product_price_model_indices.generic.append(pm_ix)

        if 'cost_price_models' in d:
            #            self.price_config.cost_price_models = [PriceModel(pm_dict) for pm_dict in d.cost_price_models]
            for pm_ix, pm_dict in enumerate(d.cost_price_models):
                if "name" not in pm_dict or ("event_name" not in pm_dict and "event_names" not in pm_dict):
                    # raise ValueError(f"Both 'name' and 'event_name' (or 'event_names') must be present in "
                    #                  f"cost price model config in {self.name}.")
                    print(f"Both 'name' and 'event_name' (or 'event_names') must be present in "
                          f"cost price model config in {self.name}. Not including cost price model.")
                    continue

                self.price_config.cost_price_models.append(PriceModel(pm_dict))
                event_names = []
                if "event_name" in pm_dict:
                    event_names = [pm_dict.event_name]
                elif "event_names" in pm_dict:
                    event_names = pm_dict.event_names
                if event_names:
                    for event_name in event_names:
                        if event_name in self.price_config.cost_price_mappings:
                            self.price_config.cost_price_mappings[event_name].append(pm_ix)
                        else:
                            print(f"Warning: '{event_name}' event referred to in cost "
                                  f"price model config does not exist in {self.name}.")

        if 'financial_events' in d:
            self.price_config
            for fe_ix, fe in enumerate(d.financial_events):
                fe.event_name = fe.name
                pm = PriceModel(fe)
                self.price_config.cost_price_models.append(pm)
                if fe.event_name in self.price_config.cost_price_mappings:
                    self.price_config.cost_price_mappings[fe.event_name].append(
                        len(self.price_config.cost_price_models)-1)
                print(fe.name)

    # Must be called after _setup_events() and _setup_price_config()
    def _setup_crop_rate_defs(self):
        crop_rate_defs = DotWiz()
        crop_rate_defs.output_rates = self.growth_model.get_output_rates()
        crop_rate_defs.propagation_product_rates = self.growth_model.get_propagation_product_rates()
        crop_rate_defs.events = {}
        for event_type, events in self.all_events.items():
            for event in events:
                prs, eors = self.growth_model.get_event_output_rates(event.name)
                cost_item_units = [self.price_config.cost_price_models[ix].denominator_unit
                                   for ix in self.price_config.cost_price_mappings[event.name]]
                crop_rate_defs.events[event.name] = {
                    'product_rates': prs,
                    'event_output_rates': eors,
                    'cost_item_units': cost_item_units
                }

        self.crop_rate_defs = crop_rate_defs

    def _build_product_price_mappings(self):
        # Now build the product price mappings.
        # Need to use the product column to find the best price model.
        for event_type, events in self.all_events.items():
            for event in events:
                if event.name in self.crop_rate_defs.events:
                    product_rates = self.crop_rate_defs.events[event.name].product_rates
                    new_price_mappings = []
                    for pr in product_rates:
                        # Find the index of a product price model that matches the product rate, or create a
                        # zero price model for it in self.price_config.relevant_product_price_model_indices.default
                        # and use that.
                        new_price_mappings.append(self._get_product_price_model_index(event.name, pr))

                    self.price_config.product_price_mappings[event.name] = new_price_mappings
                else:
                    raise ValueError(f"Expecting {event.name} in Crop.crop_rate_defs.events for crop {self.crop.name}")

    def _get_product_price_model_index(self, event_name, pr):
        # Looks up the price_config.relevant_product_price_model_indices to find a matching price model for pr.
        # If one is not found, a zero price model is added and its index is returned.
        rppmis = self.price_config.relevant_product_price_model_indices
        if event_name in rppmis:
            for price_model_index in rppmis[event_name]:
                pm = self.price_config.product_price_models[price_model_index]
                # Each of these pms was defined in the config as relevant to this event.
                # If a pm matches here, use it.
                if pm.denominator_unit == pr.unit:
                    return price_model_index

        # If not returned here, check 'generic'.
        for price_model_index in rppmis.generic:
            pm = self.price_config.product_price_models[price_model_index]
            # Each of these pms was defined in the config as relevant to this event.
            # If a pm matches here, use it.
            if pm.denominator_unit == pr.unit:
                return price_model_index

        # If not returned here, check in default.
        for price_model_index in rppmis.default:
            pm = self.price_config.product_price_models[price_model_index]
            # Each of these pms was defined in the config as relevant to this event.
            # If a pm matches here, use it.
            if pm.denominator_unit == pr.unit:
                return price_model_index

        # If not returned here, add a zero price model and return index.
        pm = PriceModel(pr.unit.measurable + " Income", unit=global_helpers.dollar, denominator_unit=pr.unit)
        price_model_index = len(self.price_config.product_price_models)
        self.price_config.product_price_models.append(pm)
        rppmis.default.append(price_model_index)
        return price_model_index

    def get_cost_price_models(self, event_name):
        event_mapping = self.price_config.cost_price_mappings[event_name]
        if len(event_mapping) == 0:
            return []
        return self.price_config.cost_price_models[event_mapping]

    def get_product_price_model(self, denominator_unit):
        matching_pms = [pm for pm in self.price_config.product_price_models if pm.denominator_unit == denominator_unit]
        if len(matching_pms) == 1:
            return matching_pms[0]
        elif len(matching_pms) == 0:
            return None
        else:
            raise ValueError(f"More than one product price model matches provided" 
                             f"denominator unit: {denominator_unit} for crop {self.name}.")

File: Matlab Code/Classes/Crop.m
classdef Crop < handle
   
    % A Crop contains data relevant to a particular crop including the
    % CropCategory, which defines the events in the crops life and the
    % kinds of products. The crop has a colour associated with it which is
    % used for graphs and images. It has a name. It has a growth model,
    % which defines the state of the crop and how the state should change.
    % It has a set of price models for the costs and for the products. Some
    % of these things will be dependent on others. For example, the growth
    % model will depend on the category.
    
    properties
        % the name, a string.
        name = '';
        
        % The colour, an RGB triple
        colour = [1 0 0];

        % FinancialEvents are those that have no transition function
        % because they were added by the user. They are only to define when
        % the farmer should incur costs.
        financialEvents = ImagineEvent.empty(1, 0);
        
    end
    
    properties (Dependent)
        
        % The category. A CropCategory object. Comes from the private
        % categories list.
        category

        % The growth model. A GrowthModel object.
        growthModel
        
        
        % The name of the category. Can set this to swap between saved
        % categories.
        categoryChoice
        
        % A list of products the crop could produce. Array of Products.
  %      products        
        
        % A list of possible costs that could be incurred. Array of
        % CostItems.
   %     costItems
        
    end
    
    properties (Access = private)
        
        categoryIndex = 0;
        categories = CropCategory.setupCategories;

    end
    
    methods (Static)
    
        function obj = loadobj(obj)
            % Just want to update the category. Resetting the choice should
            % do it.
            obj.categoryChoice = obj.categoryChoice;
            
            
            obj.growthModel.name
            
            
        end
    end
    
    methods
       
        function cropObj = Crop(nameIn, gmIn, categoryIn, colourIn, financialEventsIn)
            if nargin > 0
                cropObj.name = nameIn;
             %   cropObj.coreEvents = coreEventsIn;
                cropObj.growthModel = gmIn;
                cropObj.category = categoryIn;
             %   cropObj.products = productsIn;
             %   cropObj.costItems = costItemsIn;
                cropObj.colour = colourIn;
                cropObj.financialEvents = financialEventsIn;
            end
        end % end constructor
           
        % Gets the current category from the saved categories list via the
        % private categoryName
        function cat = get.category(obj)
            
            if obj.categoryIndex
                cat = obj.categories(obj.categoryIndex); 
            else
                cat = CropCategory.empty(1, 0);
            end
        end
 
        % To set the category object, we add it to the list, or replace it
        % if it exists, and set the categoryIndex appropriately.
        function obj = set.category(obj, cat)
            
            % Does the a category with the same name already exist? 
            ix = find(strcmp({obj.categories.name}, cat.name), 1);
            if isempty(ix)
               obj.categories(end + 1) = cat;
               obj.categoryIndex = length(obj.categories);
            else
                obj.categories(obj.categoryIndex) = cat;
            end
        end
        
        % Returns the saved category, or an empty array if it doesn't
        % exist.
        function cat = getSavedCategory(obj, catName)
            ix = find(strcmp({obj.categories.name}, catName));
            if isempty(ix)
                cat = [];
            else
               cat = obj.categories(ix); 
            end            
        end
        
        % Gets the name of the current category.
        function catChoice = get.categoryChoice(obj)
            if obj.categoryIndex
                catChoice = obj.category.name;
            else
               catChoice = ''; 
            end
        end
        
        % Swaps between saved categories if it exists. However, will add
        % from CropCategory constructor if possible.
        function crop = set.categoryChoice(crop, catChoice)
           
            cat = CropCategory(catChoice);
            if isempty(cat)
                disp('Unable to chagne categories in crop. Supplied category is not a valid choice.');
                return
            end
            
            % Find the category in the list if it exists.
            ix = find(strcmp({crop.categories.name}, catChoice), 1);
            
            if ~isempty(ix)
                % Then set the new category
                crop.categoryIndex = ix;
            else
                crop.category = cat;
            end
                        
        end % end set.categoryChoice
        
        % Gets the growthModel from the category.
        function gm = get.growthModel(obj)
            gm = obj.category.growthModel;
        end
                
        % Use the growthModel.growthModelEvents and the financialEvents to
        % get the priceModels for these events. Make sure there are no
        % repeats.
        % type is a string of 'Income' or 'Cost' and determines the type of
        % growthModels returned.
        function pms = getUniquePriceModelDefinitions(obj, type)
    
            pms = PriceModel.empty(1,0);
            
            if strcmp(type, 'Income')

                priceModels = obj.growthModel.productPriceModels;
                
            elseif strcmp(type, 'Cost')
                gmes = obj.growthModel.growthModelEvents;
                costPriceModels = [gmes.costPriceModel];
          %      extraCostPriceModels = PriceModel.empty(1, 0);
          %      for i = 1:length(gmes)
          %          extraCostPriceModels = [extraCostPriceModels gmes(i).extraCostPriceModels ];                
          %      end
                financialPriceModels = [obj.financialEvents.costPriceModel];
                priceModels = [costPriceModels, financialPriceModels];
            else
                error('Must pass a type argument. Valid types are ''Income'' and ''Cost''.');
            end
            
            % For each priceModel, check if it is already in pms.
            % If not, add it.
            for i = 1:length(priceModels)  
                alreadyIn = false;
                for j = 1:length(pms)
                    if PriceModel.definitionMatches(pms(j), priceModels(i))
                       alreadyIn = true;
                       break;
                    end
                end
                if ~alreadyIn
                    pms(end + 1) = priceModels(i);                    
                end
            end
            
            
        end % end getUniquePriceModelDefinitions
       
        % This function merges the properties of newObj into the properties
        % of origObj. The objects must be duplicableHandles and also must
        % be of the same concrete class. Where a property is a
        % duplicableObject, it absorbs newObj's property into the origObj's
        % property.
        %
        % NOTE! This method does not maintain the handles of children
        % objects. Therefore this method helps maintain the Crop's handle,
        % but it does not maintain the handles to growthModel or category
        % etc. Therefore the programmer should maintain references to the
        % sub objects through the Crop, ie store the cropObject, then get
        % it's growthModel. Don't store the handle to the growthModel.
        function absorb(origObj, newObj)
        
            if strcmp(class(newObj), class(origObj))

                mc = metaclass(origObj);
                propertyList = mc.Properties;

                for i = 1:length(propertyList)
                    propName = propertyList{i}.Name;
                    if ~(propertyList{i}.Dependent || propertyList{i}.Abstract || propertyList{i}.Constant || propertyList{i}.Transient)
                        origObj.(propName) = newObj.(propName);
                    end                                            
                end

            else
               error('Absorb method can only be called on objects of exactly the same type.'); 
            end
           
        end 
    end
    

    
    methods (Static)
       
        % checks that the object passed is really a crop.
        % It can also be a list of crops. List of booleans are returned.
        function valid = isValid(crops)
            valid = zeros(1, length(crops));
            for i = 1:length(crops)
                c = crops(i);
                valid(i) = isa(c, 'Crop');
                if valid(i)
                    valid(i) = all([ ischar(c.name), ...
                                    isnumeric(c.colour), ...
                                    size(c.colour) == [3, 1] ]); %...
                                  %  CropCategory.isValid(c.category), ...
                                  %  GrowthModel.isValid(c.growthModel), ...
                                  %  Product.isValid(c.products), ...
                                   % CostItem.isValid(c.costItems)]);
                                  
                end
                
                if ~valid
                    valid = 1
                    disp('Overriding Crop to valid');
                end
                
            end    
            
        end % end Crop.isValid
        
    end


    
    
    
end